tests/Util/Examples/ExampleRunnerTest.cs

```diff
@@ -0,0 +1,170 @@
+            );
+
+            // Verify that a KeyNotFoundException is thrown if the specified example
+            // is not found.
+            CaptureConsoleIOAndExecute(
+                delegate ()
+                {
+                    Assert.Throws(typeof(KeyNotFoundException),
+                        delegate ()
+                        {
+                            runner.Run("V0.NoSuchExample", client, new string[] { });
+                        }
+                    );
+                }, null
+            );
+
+            // Verify that a MissingMethodException is thrown if the specified example
+            // does not have a Run() method.
+            CaptureConsoleIOAndExecute(
+                delegate ()
+                {
+                    Assert.Throws(typeof(MissingMethodException),
+                        delegate ()
+                        {
+                            runner.Run(NoRunTestExample.TEST_EXAMPLE_NAME, client,
+                                new string[] { });
+                        }
+                    );
+                }, null
+            );
+
+            // Verify that a ArgumentException is thrown if an unknown flag is passed.
+            CaptureConsoleIOAndExecute(
+                delegate ()
+                {
+                    Assert.Throws(typeof(ArgumentException),
+                        delegate ()
+                        {
+                            runner.Run(TestExample.TEST_EXAMPLE_NAME, client,
+                                new string[] { "--foo=bar" });
+                        }
+                    );
+                }, null
+            );
+        }
+
+        /// <summary>
+        /// Captures the console input + output and execute a method.
+        /// </summary>
+        /// <param name="functionToRun">The function to run. This function should not read any
+        /// input from the console.</param>
+        /// <param name="afterRunCallback">The callback to be invoked after
+        /// <paramref name="functionToRun"/> completes. The console output is passed as an
+        /// argument to the action.</param>
+        private void CaptureConsoleIOAndExecute(TestDelegate functionToRun,
+            Action<string> afterRunCallback)
+        {
+            StringWriter writer = new StringWriter();
+            TextWriter oldWriter = null;
+            TextReader oldReader = null;
+            StringReader reader = new StringReader("\r\n");
+
+            try
+            {
+                // Capture the console input and output.
+                oldWriter = Console.Out;
+                Console.SetOut(writer);
+                oldReader = Console.In;
+                Console.SetIn(reader);
+
+                functionToRun?.Invoke();
+                afterRunCallback?.Invoke(writer.ToString());
+            }
+            finally
+            {
+                // Restore the console input and output.
+                Console.SetIn(oldReader);
```

Comment:
I may be misreading this, but it looks like we're trying to reset the input as Console.In, and reset the output as Console.Out and after we've run and recorded the function.  The issue i'm seeing is that we're initializing the oldWriter and oldReader as null, then setting them inside the try.  If the Console.SetOut() method fails out, the oldReader will never be set to Console.In, and the finally will try to set Console.SetIn(null), which I'm assuming would fail.

Instead, can we initialize the oldWriter/oldReader variables at the declaration in lines 147/148?  This should prevent that issue.

Reply:
Done.